arch: arc: let zephyr run in normal mode of arc

vonhust opened the pull request:
This PR is initial implementation of some kind of TEE support for ARC.

As the TEE design/framework in zephyr matures,  more work will be needed. 

It includes necessary commits to let zephyr run in normal mode of
arc processors equipped with SecureShield.

In this PR, zephyr can be built into:
1. imgage for secure mode (CONFIG_ARC_SECURE_FIRMWARE <- TRUSTED_EXECUTION_SECURE)
   * full access to system resource
   * full zephyr features are supported,  USERSPACE, stack protection
2. image for normal mode (CONFIG_ARC_NORMAL_FIRMWARE)
  * no mpu support in normal mode, i.e., no USERSPACE
  * to run this normal image,  it requires a secure bootloader or secure zephyr application to
     boot, as when processor is reset, it's secure mode.

To show how normal zephyr application works with secure zephyr applicaiton, an example is provided in samples/boards/arc_secure_services,  more details, pls go to the README.rst of this example.

ruuddw commented on arch/arc/core/fault_s.S:
```diff
@@ -95,7 +95,11 @@ _exc_entry:
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
 	lr r0,[_ARC_V2_ERSEC_STAT]
+#else
+	mov r0, 0
```
I think it's actually ok to read ERSEC_STAT from Normal mode as well - fields without access will be 0 (IOW/RAZ)

vonhust replied:
read is ok, but no write. It will make codes simpler

ruuddw commented on arch/arc/core/prep_c.c:
```diff
@@ -88,8 +88,13 @@ static void invalidate_dcache(void)
 {
 #ifdef CONFIG_ARC_HAS_SECURE
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
+/* only when processor in secure mode, the vector base should be
```
Why? Normal mode has it's own vector table, shouldn't that be used (and updated) in Normal mode?

ruuddw commented on arch/arc/core/swap.S:
```diff
@@ -92,7 +92,11 @@ SECTION_FUNC(TEXT, __swap)
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
 	lr r3, [_ARC_V2_SEC_STAT]
+#else
+	mov r3, 0
```
See earlier remark on ERSEC_STAT, I think it's ok to access from Normal mode as well.

ruuddw commented on arch/arc/core/thread.c:
```diff
@@ -115,7 +115,11 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t *stack,
 #endif
 
 #ifdef CONFIG_ARC_HAS_SECURE
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
```
Is reading the current contents best way to initialize the SEC_STAT in the initial context? Can't it be some constant value?

vonhust replied:
It could be  a constan value.  I will consider to change it later.

ruuddw commented on arch/arc/core/vector_table.c:
```diff
@@ -65,3 +65,31 @@ struct vector_table _VectorTable _GENERIC_SECTION(.exc_vector_table) = {
 	0
 };
 
+#ifdef CONFIG_ARC_NORMAL_FIRMWARE
```
Assume this is for the temporary secure initialization code, it's not really 'logical' to declare secure stuff in a 'NORMAL_FIRMWARE' conditional section. (maybe add a 'TODO' note)

ruuddw commented on arch/arc/include/swap_macros.h:
```diff
@@ -44,10 +44,17 @@ extern "C" {
 
 #ifdef CONFIG_USERSPACE
 #ifdef CONFIG_ARC_HAS_SECURE
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
```
not sure we cover all combination of options, looks like HAS_SECURE implies we also have USERSPACE - no support for Secure but no userspace?

vonhust replied:
if no userspace, no need to save and restore  USER/KERNEL_SP realted aux regs.

ruuddw commented on arch/arc/include/swap_macros.h:
```diff
@@ -247,7 +261,7 @@ extern "C" {
  */
 .macro _load_stack_check_regs
-#ifdef CONFIG_ARC_HAS_SECURE
+#if defined(CONFIG_ARC_HAS_SECURE) && defined(CONFIG_ARC_SECURE_FIRMWARE)
```
Missing case for HAS_SECURE and NORMAL_FIRMWARE?

vonhust replied:
fixed

ruuddw commented on arch/arc/include/v2/irq.h:
```diff
@@ -52,8 +52,12 @@ static ALWAYS_INLINE void _irq_setup(void)
 
 	k_cpu_sleep_mode = _ARC_V2_WAKE_IRQ_LEVEL;
+#ifdef CONFIG_ARC_NORMAL_FIRMWARE
+	/* \todo use sjli instruction */
```
not sure an sjli call would be required here, this setup is only done once and could be done by Secure firmware before switching to Normal Zephyr.

ruuddw commented on boards/arc/nsim_em/nsim_sem_normal.yaml:
```diff
@@ -0,0 +1,11 @@
+identifier: nsim_sem
```
copy/paste typo: this should be nsim_sem_normal

ruuddw commented on drivers/interrupt_controller/arcv2_irq_unit.c:
```diff
@@ -162,8 +162,11 @@ static int _arc_v2_irq_unit_resume(struct device *dev)
 	}
-
+#ifdef CONFIG_ARC_NORMAL_FIRMWARE
+	/* \todo use sjli instruction to access irq_ctrl */
```
maybe suspend/resume functionality should have different overall implementation, Secure firmware should probably also support suspend/resume for this to work.

ruuddw commented on include/arch/arc/v2/arcv2_irq_unit.h:
```diff
@@ -98,7 +98,7 @@ static ALWAYS_INLINE
 {
 	_arc_v2_aux_reg_write(_ARC_V2_IRQ_SELECT, irq);
-#ifdef CONFIG_ARC_HAS_SECURE
+#if defined(CONFIG_ARC_HAS_SECURE) && defined(CONFIG_ARC_SECURE_FIRMWARE)
```
All interrupts either Secure or Normal is probably too strong, some should be allocated to Secure and some to Normal depending on application. For now ok, but add a note (or issue) to support configuration/allocation of interrupts to Secure or Normal.

vonhust replied:
It requires a mechinism to tell which interrupts are secure, which are normal

nashif commented:
@vonhust @ruuddw can you please rebase and drive to closure? thanks.

nashif commented:
@ruuddw are you ok with this?

nashif commented:
> @nashif still work in progress, I've marked it DNM for now.

ok

dbkinder commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,17 @@
+.. _arc_secure_service:
```
Ping me when there's documentation to review...

vonhust replied:
pls review the documentation.

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -135,9 +135,59 @@ config CODE_DENSITY
+	default y if TRUSTED_EXECUTION_NONSECURE
+	help
+	  This option indicates that we are building a Zephyr image that
+	  is intended to execute in normal state. Execution of this
```
"normal state" -> normal mode

vonhust replied:
fixed

ruuddw commented on arch/arc/core/fault_s.S:
```diff
@@ -112,29 +116,28 @@ _exc_return:
+
+	mov r3, 0
+	st r3, [exc_nest_count]
+	/* pretend lowest priority interrupt happened to use common handler
```
Since this requires a sjli call in Normal mode, maybe better now to not use re-use the common handler?
Guess this code is not in a timing critical path, but avoiding a slji would be good. It's a powerfull/dangerous sjli call so should not be use-able by other normal code (implementation of the sjli needs to check the caller)

ruuddw commented on arch/arc/core/irq_manage.c:
```diff
@@ -83,6 +83,16 @@ void _irq_priority_set(unsigned int irq, unsigned int prio, u32_t flags)
 
 	__ASSERT(prio < CONFIG_NUM_IRQ_PRIO_LEVELS,
 		 "invalid priority %d for irq %d", prio, irq);
+/* 0 -> CONFIG_NUM_IRQ_PRIO_LEVELS allocted to secure world
```
is this split required? This enforces secure priorities are higher than normal priorities. And, it requires hardware configuration with multiple priority levels (guess most ARC configs would have that anyway).

vonhust replied:
I think it's required, or the normal mode can attack the secure mode through interrupt. With this assumption, the whole system is more secure and simpler, e.g. the cases of bounce of interrupts between secure and normal.  It's just something like fast irq's priority is higher then regular irq

ruuddw replied:
Interrupt priorities are typically assigned based on real-time response requirements. That can be different from security concerns. Fast vs regular IRQ's is about response times, so different.

ruuddw commented on arch/arc/core/regular_irq.S:
```diff
@@ -178,6 +177,11 @@ _rirq_common_interrupt_swap:
 
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
+	lr r0, [_ARC_V2_SEC_STAT]
+	bset r0, r0, _ARC_V2_SEC_STAT_IRM_BIT
```
not all interrupts are secure, or return to secure mode - why set IRM?

vonhust replied:
here, rtie is used to do a dispatch.  The interruptted context could be in normal mode and the target thread must be in secure mode. So the  IRM must be set to the correct value.

ruuddw replied:
Why must target thread be in secure mode?

ruuddw commented on arch/arc/core/secureshield/arc_secure.S:
```diff
@@ -0,0 +1,122 @@
+SECTION_FUNC(TEXT, _arc_do_secure_call)
+	/* r0-r5: arg1-arg6, r6 is call id */
+	/* the call id should be checked */
+	/* disable normal interrupt happened when processor in secure mode ? */
```
preferably normal interrupts are allowed during secure calls.

vonhust replied:
normal mode can attack the secure mode through interrupt. So for more secure system, do not let normal world preempt the secure world

ruuddw commented on arch/arc/core/secureshield/arc_secure.S:
```diff
@@ -0,0 +1,122 @@
+	breq r6, ARC_S_CALL_CLRI, _s_clri
+	breq r6, ARC_S_CALL_SETI, _s_seti
+	push_s blink
+	mov	blink, arc_s_call_table
```
Why is a SW dispatch table used instead of using multiple sjli calls and hw dispatching?

ruuddw commented on arch/arc/core/secureshield/arc_secure.S:
```diff
@@ -0,0 +1,122 @@
+
+SECTION_FUNC(TEXT, _arc_do_secure_call)
+	/* r0-r5: arg1-arg6, r6 is call id */
+	/* the call id should be checked */
```
not just call id, also other arguments and - depending on the call id/secure system call feature requested - maybe even whether the caller is allowed to call the function.

vonhust replied:
it should be left to the implementation of secure system call handler.

ruuddw commented on arch/arc/core/secureshield/secure_sys_services.c:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @param intno, the interrupt to be allocated to normal world
+ *
+ * By default, most interrupts are configured to be secure in initialization.
```
I'd expect that Secure mode at system initialization defines certain interrupts as Normal, and rest as Secure. Not sure Normal mode needs this call.

vonhust replied:
There may be some dynamic cases.  If the interrupt allocation is static, yes, no need such a call.

ruuddw commented on arch/arc/core/swap.S:
```diff
@@ -188,22 +191,22 @@ _swap_return_from_firq:
+	mov r0, _ARC_V2_AUX_IRQ_ACT
+	mov r1, r3
+	mov r6, ARC_S_CALL_AUX_WRITE
+	sjli SJLI_CALL_ARC_SECURE
```
is it actually necessary to set IRQ_ACT before doing a 'fake' rtie? Would be good if it can be avoided.

vonhust replied:
currently, no better way except we don't use rtie instruction. If not sec IRQ_ACT,  the rtie may call the exception return flow

ruuddw commented on boards/arc/em_starterkit/arc_mpu_regions.c:
```diff
@@ -50,11 +52,13 @@ static struct arc_mpu_region mpu_regions[] = {
 			AUX_MPU_RDP_KE | AUX_MPU_RDP_UE),
 #endif
 #endif /* ARC_MPU_VER == 3 */
+#ifndef CONFIG_ARC_SECURE_FIRMWARE
```
ultimately, peripherals should be split over Normal and Secure mode. Maybe normal and secure board dts files can be used for that.

vonhust replied:
It depends on the application. Currently, I just let the peripherals shared between secure and normal mode

ruuddw commented on samples/boards/arc_secure_services/src/main.c:
```diff
@@ -0,0 +1,59 @@
+#define SLEEPTIME 1000
+
+
+void threadA(void *dummy1, void *dummy2, void *dummy3)
```
is a thread required in Secure mode? We don't support mixed secure/normal threads (yet...), the Secure part is more like a library and initial entry point before handing over to Normal mode running full Zephyr.

vonhust replied:
It's a complicate case. It thread is not required, just disable the multithreading in Zephry. If we can let two zephyr run together, it will be simple for baremetal case.

ruuddw commented on arch/arc/Kconfig:
```diff
@@ -135,9 +135,59 @@ config CODE_DENSITY
+	depends on ARC_SECURE_FIRMWARE
+	default 8
+	help
+	   The size of sjli (Secure Jump and Linux Indexed) table. The
```
interesting typo, can jump to Linux with a sjli call ;-)

vonhust replied:
fixed

vonhust commented on arch/arc/core/reset.S:
```diff
@@ -51,8 +189,10 @@ SECTION_FUNC(TEXT,__start)
 	kflag r0
 
 	/* interrupt related init */
+#ifndef CONFIG_ARC_NORMAL_FIRMWARE
```
done

nashif commented:
@ruuddw  @vonhust  still relevant, if yes, please update and rebase?

vonhust commented:
@nashif @ruuddw  I just rebase ARC TEE support on zephyr latest code. We can start the review again.
But i met a small problem reported by CI,  Dependcy loop issuse in kconfig. This issue also made "make htmldocs". I already took a look at this issue, didn't find a quick fix. Any ideas or hints?

Commit:
arch: arc: add initial support of ARC TEE

* it's based on ARC SecureShield
* add basic secure service in arch/arc/core/secureshield
* necesssary changes in arch level
   * thread switch
   * irq/exception handling
   * initialization
* add secure time support

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
boards: emsk: add initial support of normal/non-secure application

* Non-secure/normal application: em_starterkit_em7d_normal
* secure application: em_starterkit_em7d

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
boards: nsim: add initial support of normal/non-secure application

normal/non-secure application: nsim_sem_normal

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
samples: add sample to show how ARC TEE works

* this is a simple sample to show how
 secure applicaiton and non-secure application work
 together. More details are in README.rst

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

vonhust commented:
@ruuddw @abrodkin  I refactor this PR for 2.0  release. Although still very primitive,  it's a good start point for ARC TEE support

ioannisg commented:
@vonhust is this PR abandoned?
If not, shall we move it to 2.1 release?

vonhust commented:
> @vonhust is this PR abandoned?
> If not, shall we move it to 2.1 release?

It not abandoned,  i am trying to make it 2.0 in the left 2-3 days

ioannisg commented:
> 
> 
> > @vonhust is this PR abandoned?
> > If not, shall we move it to 2.1 release?
> 
> It not abandoned, i am trying to make it 2.0 in the left 2-3 days

Oh cool! @vonhust So sorry,
I saw the conflicts and the PR "age" but didn't see you actually are working on it.

ioannisg reviewed (commented):
I've only checked the generic arch code - that looks ok to me

dbkinder commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+Normal application
+^^^^^^^^^^^^^^^^^^
+
+Currently, in normal application, MPU is not accessible, so no user
```
Change to:
```
In the normal application, the MPU is not accessible, so there is
no user space and MPU-based stack checking.
```

dbkinder commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+Currently, in normal application, MPU is not accessible, so no user
+space and mpu-based stack checking.
+
+Here,take :ref:'dining-philosophers-sample' as an example for normal
```
Change to:
```
We'll use the  :ref:'dining-philosophers-sample' as an example
for the normal application.
```

dbkinder commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+
+To run this example, there are two ways.
+
+* By bootloader
```
Change this to:
```
Run using the bootloader
   The bootloader should load the secure and normal application into the correct place,
   then jump to the entry of the secure application.

Run using the debugger (recommended)
   Use the gdb debugger to load and run the two applications.
```

vonhust replied:
Fixed

dbkinder commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+
+Currently, it's recommended to use to gdb debugger to load and run two applications.
+
+For em starter kit, please run the following commands.
```
delete ``please``

vonhust replied:
Fixed

dbkinder commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+    $ monitor load_image <the path of normal application elf file>
+    $ c
+
+For nsim sem, you need two consoles, one for output, one for debug.
```
Change to:
```
For nsim sem, you need two consoles: one for
application output, and one for the debugger.
```

vonhust replied:
Fixed

dbkinder commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+    $ cd samples/boards/arc_secure_services
+    $ west debugserver
+
+In the console for debug:
```
``for the debugger``

vonhust replied:
Fixed

dbkinder commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+    # load normal application
+    $ load <the path of normal application elf file>
+    # load secure application
+    $ load
```
do you need to specify the path of the secure application, or is that on line 120?

vonhust replied:
<the path of normal application elf file> means any zephyr image built in line 73, i will make it more clarified

vonhust commented on arch/arc/core/reset.S:
```diff
@@ -27,6 +27,143 @@ GDATA(_main_stack)
 GTEXT(__reset)
 GTEXT(__start)
 
+#ifdef CONFIG_ARC_NORMAL_FIRMWARE
```
It's removed in latest codes

ruuddw reviewed (approved):
There's remaining work to do, but this is a good start for people that want to experiment with this on ARC.
Let's get it into v2.0.

ioannisg commented:
> 
> 
> There's remaining work to do, but this is a good start for people that want to experiment with this on ARC.
> Let's get it into v2.0.

@vonhust have you addressed the latest feedback by @dbkinder ?

vonhust commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+  ``arc_mpu_regions.c``. Half of ram and rom is allocated to secure world,
+  the other half is allocated to normal world.
+
+* the memory which is not allocated to secure application is allocated to
```
fixed

vonhust commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+In this sample:
+
+* secure application will be in the secure memory space defined in
+  ``arc_mpu_regions.c``. Half of ram and rom is allocated to secure world,
```
fixed

vonhust commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+
+In this sample:
+
+* secure application will be in the secure memory space defined in
```
fixed

vonhust commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+
+This sample implements a simple secure service based on ARC SecureShield to
+demonstrate how a secure zephyr application runs together with a normal
+application, e.g. normal zephyr.
```
Fixed

vonhust commented on samples/boards/arc_secure_services/README.rst:
```diff
@@ -0,0 +1,126 @@
+********
+
+This sample implements a simple secure service based on ARC SecureShield to
+demonstrate how a secure zephyr application runs together with a normal
```
Fixed

dbkinder reviewed (dismissed):
I see you mention my comments were "fixed", but I don't see the changes actually pushed.

vonhust commented:
> , but I don't see the changes actually pushed.

I made a force push, could you check that again?

ioannisg commented:
Seems to me the fixes were pushed (but might be missing something @dkinder).

vonhust commented:
better to merge after PR #18079 and #18142

ioannisg commented:
> 
> 
> better to merge after PR #18079 and #18142

@vonhust please resolve the conflicts, after merging the above PRs.
Note that the review needs to conclude today, if this is to get to 2.0 release.

vonhust commented:
> > better to merge after PR #18079 and #18142
> 
> @vonhust please resolve the conflicts, after merging the above PRs.
> Note that the review needs to conclude today, if this is to get to 2.0 release.

confilct fixed, ok to merge

ioannisg commented:
> 
> 
> > > better to merge after PR #18079 and #18142
> > 
> > 
> > @vonhust please resolve the conflicts, after merging the above PRs.
> > Note that the review needs to conclude today, if this is to get to 2.0 release.
> 
> confilct fixed, ok to merge

There is still a request-for-changes by @dbkinder.
@vonhust can this PR go into 2.0.0 regardless of what happens with #17767?

vonhust commented:
> > > > better to merge after PR #18079 and #18142
> > > 
> > > 
> > > @vonhust please resolve the conflicts, after merging the above PRs.
> > > Note that the review needs to conclude today, if this is to get to 2.0 release.
> > 
> > 
> > confilct fixed, ok to merge
> 
> There is still a request-for-changes by @dbkinder.
> @vonhust can this PR go into 2.0.0 regardless of what happens with #17767?

There is no relation between this PR and #17767. I just dismissed dbkinder's request as i addressed his feedbacks. Let's go for merge

ioannisg commented:
> 
> 
> > > > > better to merge after PR #18079 and #18142
> > > > 
> > > > 
> > > > @vonhust please resolve the conflicts, after merging the above PRs.
> > > > Note that the review needs to conclude today, if this is to get to 2.0 release.
> > > 
> > > 
> > > confilct fixed, ok to merge
> > 
> > 
> > There is still a request-for-changes by @dbkinder.
> > @vonhust can this PR go into 2.0.0 regardless of what happens with #17767?
> 
> There is no relation between this PR and #17767. I just dismissed dbkinder's request as i addressed his feedbacks. Let's go for merge

Yes. @dbkinder if there are still imperfections in documentation, we can definitely address them during the stabilization period.
